text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+  gives them.  `EXIT_SUCCESS` is always 0.  `EXIT_FAILURE` has the same
+  value as the C library gives it, _unless_ the C library gives it the
+  value 1, in which case 2 is used instead.
+
+* All of the impls of `Termination` in the stdlib are guaranteed to
+  use only `EXIT_SUCCESS` and `EXIT_FAILURE`, with one exception:
+
+* There is a type, provisionally called `TermStatus`, which is a
+  newtype over `i32`; on Unix (but not on Windows), creating one from
+  a value outside the range 0 ... 255 will panic.  It implements
+  `Termination`, passing its value to `exit` and not printing any
+  diagnostics.  Using this type, you can generate specific exit codes
+  when appropriate, without having to avoid using `?` in `main`.
+
+  (It can't be called `ExitStatus` because that name is already
+  taken for the return type of `std::process::Command::status`.)
+
+[exit.3]: http://www.cplusplus.com/reference/cstdlib/exit/
+[grep.1]: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html
+[automake-tests]: https://www.gnu.org/software/automake/manual/html_node/Scripts_002dbased-Testsuites.html
+[sysexits]: https://www.freebsd.org/cgi/man.cgi?query=sysexits
+
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+This should be taught alongside the `?` operator and error handling in
+general.  The stock `Termination` impls in libstd mean that simple
+programs that can fail don't need to do anything special:
+
+``` rust
+fn main() -> Result<(), io::Error> {
+    let mut stdin = io::stdin();
+    let mut raw_stdout = io::stdout();
+    let mut stdout = raw_stdout.lock();
+    for line in stdin.lock().lines() {
+        stdout.write(line?.trim().as_bytes())?;
+        stdout.write(b"\n")?;
+    }
+    stdout.flush()
+}
+```
+
+Doctest examples can freely use `?` with no extra boilerplate;
+`#[test]` examples may need their boilerplate adjusted.
+
+More complex programs, with their own error types, still get
+platform-consistent exit status behavior for free, as long as they
+implement `Error`.  The Book should describe `Termination` to explain
+_how_ `Result`s returned from `main` turn into error messages and exit
+statuses, but as a thing that most programs will not need to deal with
+directly.
+
+Tutorial examples should probably still begin with `fn main() -> ()`
+until the tutorial gets to the point where it starts explaining why
+`panic!` and `unwrap` are not for "normal errors".
+
+Discussion of `TermStatus` should be reserved for an advanced-topics
+section talking about interoperation with the Unix command-line
+ecosystem.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Generalizing the return type of `main` complicates libstd and/or the
+compiler.  It also adds an additional thing to remember when complete
+newbies to the language get to error handling.  On the other hand,
+people coming to Rust from other languages may find this _less_
+surprising than the status quo.
+
+# Alternatives
```

Comment:
Or encourage the use of templates like [quickstart](https://github.com/rusttemplates/quickstart). That's how I start all my toy and non-toy `std` apps today. Granted, that wouldn't help with doctests / documentation.